### PR TITLE
Implement Dispose pattern

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
The extension now properly supports the dispose pattern by clearing any properties and unregistering all events.

# Notes
I'm not actually sure if I implemented the dispose pattern in the same way that the BCL does.

### CODE REVIEW

File: DSharpPlus.Commands/CommandsExtension.cs [hunk 1]
```diff
@@ -162,23 +161,34 @@ public void AddCommands(IEnumerable<Type> types)
...
     {
-        this._processors.Add(processor.GetType(), processor);
-        await processor.ConfigureAsync(this);
+        if (this._processors.TryGetValue(processor.GetType(), out ICommandProcessor? existingProcessor))
+        {
+            existingProcessor.Dispose();
+        }
+
+        this._processors[processor.GetType()] = processor;
+        return processor.ConfigureAsync(this);
```
Comment (akiraveliara) [inline DSharpPlus.Commands/CommandsExtension.cs:172]: this ruins the async stacktrace, let's not

**Review by akiraveliara — CHANGES_REQUESTED** [review]
you know what i think about your formatting, but...

File: DSharpPlus.Commands/CommandsExtension.cs [hunk 2]
```diff
@@ -263,18 +266,25 @@ public async Task RefreshAsync()
...
+        foreach (ICommandProcessor processor in this._processors.Values)
+        {
+            processor.Dispose();
+        }
+
+        Commands = new Dictionary<string, Command>();
+        Processors = new Dictionary<Type, ICommandProcessor>();
+        this._commandBuilders.Clear();
+        this._processors.Clear();
+        this._checks.Clear();
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/CommandsExtension.cs:289]: Readonly fields + `[]` might create a new instance.

File: .editorconfig [hunk 3]
```diff
@@ -26,12 +26,17 @@ charset = utf-8
 ###############################
 [*.{cs,vb}]
 
+# Don't use primary constructors
+csharp_style_prefer_primary_constructors = false:silent
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply .editorconfig:30]: I think so

[the remark this answers is not in the document]
Comment (Plerx2493) [reply .editorconfig:30]: not really but i think we three agreed to it

**Review by Plerx2493 — APPROVED** [review]
LTGM

**Pull request closed without merging**